### PR TITLE
Authentication with Bandada

### LINKED ISSUE #238: Use Bandada to authenticate users
# Prerequisites

Please answer the following question for yourself before submitting a feature request.

-   [X] I checked to make sure that this request has not already been filed

# Motivation

Is your feature request related to a specific problem? Is it just a crazy idea? Tell us about it!

We have discussed adding additional authentication methods. Bandada gives us a great deal of flexibility without the overhead of  developing new authentication methods independently.

# Details

Please describe your feature request in detail.

Make sure that you have reviewed [p0tion's Contributor Guidelines](https://github.com/privacy-scaling-explorations/p0tion/blob/main/CONTRIBUTING.md)

Authentication via Bandada will provide an alternative to the existing (GitHub-only) auth method. 

The anticipated method will involve developing a Firebase custom authentication function to verify Bandada group membership . Then, Bandada groups will be created to suit common or ceremony-specific requirements. An example would be EThereum accounts with nonce>32

### PR DESCRIPTION
**Pull request by NicoSerranoP**
---
name: Authentication with Bandada
about: Allow users to login/authenticate using Bandada groups (Github or Ethereum in the future)
---

# Description

Bandada allows users to prove membership of a specific group without revealing the specific member identity (e.g. I could show I belong to the group with members only with more than 2 followers on Github without revealing my Github username). Bandada is planned to be an authentication provider in the future for Github users and Ethereum users (e.g. addresses with more than 3 transactions).

By implementing Bandada in P0tion we take advatange of the existing infrastructure and make it more scalable as the Bandada team keeps developing new features.

Fixes #238 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

-   [ ] Test A
-   [ ] Test B

# Checklist:

-   [ ] My code follows the style guidelines of this project
-   [ ] I have performed a self-review of my own code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [ ] My changes generate no new warnings
-   [ ] Any dependent changes have been merged and published in downstream modules
-   [ ] I reviewed the [code of conduct](https://github.com/privacy-scaling-explorations/p0tion/blob/main/CODE_OF_CONDUCT.md) and [contributors' guide](https://github.com/privacy-scaling-explorations/p0tion/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Review by glamperd — APPROVED** [review]
looks good

File: packages/backend/package.json [hunk 1]
```diff
@@ -76,7 +78,7 @@
         "html-entities": "^2.3.3",
         "rimraf": "^5.0.0",
         "rollup": "^3.21.6",
-        "snarkjs": "^0.6.11",
+        "snarkjs": "^0.7.2",
```
Comment (0xjei) [inline packages/backend/package.json:81]: nit: mismatching with the root `package.json` (^0.7.7)

File: packages/backend/src/functions/bandada.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import { BandadaValidateProof } from "../types/index"
+
+dotenv.config()
+
+const { BANDADA_API_URL, BANDADA_GROUP_ID } = process.env
+
+const bandadaApi = new ApiSdk(BANDADA_API_URL)
+
+export const bandadaValidateProof = functions
```
Comment (0xjei) [inline packages/backend/src/functions/bandada.ts:14]: +1

File: packages/phase2cli/src/commands/authBandada.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        identity,
+        tempGroup,
+        BANDADA_GROUP_ID,
+        signal,
+        {
+            wasmFilePath: "/home/nnico/ethereum/p0tion/packages/phase2cli/semaphore.wasm",
+            zkeyFilePath: "/home/nnico/ethereum/p0tion/packages/phase2cli/semaphore.zkey"
+        }
+    )
+    console.log(proof)
```
Comment (0xjei) [inline packages/phase2cli/src/commands/authBandada.ts:51]: neat: remove or prettify output in the console (use `theme`)

**Review by 0xjei — APPROVED** [review]
Great job, Nico! :heart_on_fire: 

I want to underline the necessity of configuring a specific group in the environment. This ensures that all ceremonies will reference that group for authentication, allowing the same identity to contribute to multiple ceremonies just proving the inclusion :100: (this avoids many small groups x ceremony with data duplication)

Indeed, this stress-test is incredibly valuable for the Bandada infrastructure to assess its scalability :100:

There are just a few nits but overall LGTM :+1:

File: packages/backend/src/functions/bandada.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            "7899194018737016222260328309937800777948677569409898603827268776967707173231",
+            "1"
+        ]
+    ]
+}
+
+dotenv.config()
+
+const { BANDADA_API_URL, BANDADA_GROUP_ID } = process.env
+
```
Comment (ctrlc03) [inline packages/backend/src/functions/bandada.ts:109]: why are these global var? could we put them inside the cloud function?

File: packages/backend/src/functions/bandada.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
+import dotenv from "dotenv"
+import * as functions from "firebase-functions"
+import { ApiSdk } from "@bandada/api-sdk"
+import { groth16 } from "snarkjs"
+import { getAuth } from "firebase-admin/auth"
+import { BandadaValidateProof, VerifiedBandadaResponse } from "../types/index"
+
+import admin from "firebase-admin"
+
+const VKEY_DATA = {
```
Comment (ctrlc03) [inline packages/backend/src/functions/bandada.ts:10]: should we move this to an util/constants file?

File: packages/phase2cli/src/commands/authBandada.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    // 5. generate a proof that the user owns the commitment.
+    spinner.text = `Generating proof of identity...`
+    spinner.start()
+    // publicSignals = [hash(externalNullifier, identityNullifier), commitment]
+    const { proof, publicSignals } = await groth16.fullProve(
+        {
+            identityTrapdoor: identity.trapdoor,
+            identityNullifier: identity.nullifier,
+            externalNullifier: BANDADA_GROUP_ID
```
Comment (ctrlc03) [inline packages/phase2cli/src/commands/authBandada.ts:64]: is it ok that the externalNullifier is always the same?

**Review by ctrlc03 — APPROVED** [review]
Great work Nico!! just left couple of comments and a question!

**Pull request merged**